presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java

```diff
@@ -112,7 +123,24 @@ private static boolean isPartitionedJoin(JoinNode joinNode)
         }
 
         @Override
-        public PlanNode visitJoin(JoinNode joinNode, RewriteContext<Void> context)
+        public PlanNode visitProject(ProjectNode node, RewriteContext<Set<VariableReferenceExpression>> context)
+        {
+            if (alwaysEnabled || !enabledForKeyFromOuterJoin) {
```

Comment:
Seems like RandomizeNullKeyInOuterJoin has two flavors and two different flags represents different things. If that is true then alwaysEnabled seems misleading. We should rename it to reflect the part it is responsible for.

Reply:
The visitProject function is to collect the output from outer joins when it's getting renamed in projection node. The check on `alwaysEnabled` is because we do not need to collect these renamed VariableReferences as we will not use it (we will enabled this optimization anyway).

Yeah, with this PR the optimization will have two modes, one is always enabled, and the other one is to enabled only when the join key is from outer join and can have many nulls. Ideally we should use a enum to set the mode, but this optimization is already used in production and just cannot do this due to compatibility.

Reply:
You can add a new session property that is an enum, and use older one as a "deprecated" session property that overrides this enum value. It will be cleaner to deal with complex if conditions. 

Look at DISTRIBUTED_JOIN as an example which was deprecated